README.md
# mysqld cookbook

Manage your mysqld servers with this cookbook.
Unlike the official [opscode mysql cookbook](https://github.com/opscode-cookbooks/mysql),
it doesn't mess with the default mysql configuration. If you do not specify anything explicitly, the
defaults of your distribution will be used.
(At least if I do not mess up - Check the
[default attributes](https://github.com/chr4/chef-mysqld/blob/master/attributes/defaults.rb),
if unsure (and file a pull request if you need to correct anything))

Features

* Defaults to OS settings unless explicitly specified otherwise
* Supports **all** my.cnf settings

## Compatibility
This cookbook is currently tested on Ubuntu 16.04. It should work on other Ubuntu/ Debian systems as well, as long as a compatible database version is used (e.g. `>= mysql-5.7.6` or `>= mariadb-10.1`).

For older database versions, it might be required to adapt some of the `my.cnf` attributes. Furthermore, the `mysqld_password` provider only works with the mentioned compatible databases. In case you need to use `mysqld_password` with older versions (e.g. `< mysql-5.7.6`), you need to version-lock this cookbook to `1.0.5`. This can be achieved by putting the following in your `Berksfile`:

```ruby
cookbook 'mysqld', '~> 1.0'
```

**Note: The support for RHEL was dropped with version `v2.2.0`, pull-requests are welcome!**

[Contributions](https://github.com/chr4/chef-mysqld#contributing) to support other systems are very
welcome!

## Requirements

You need to add the following line to your metadata.rb

    depends 'mysqld'


## Attributes

### Configuration

Everything in your my.cnf can be maintained using attributes.
Consider using the provides LWRPs (see below)

If you do not specify anything, the defaults of your os will be used.

This recipe supports **every** setting in the my.cnf.  All your settings will be merged with the
systems default, and then written to the my.cnf config file. The packages to install, the path to
my.cnf as well as the name of the service are set automatically, and can be overwritten using the
following attributes:


```ruby
node['mysqld']['my.cnf_path']
node['mysqld']['service_name']
node['mysqld']['mysql']['packages']
node['mysqld']['mariadb']['packages']
node['mysqld']['mariadb']['galera_packages'] # When using mariadb_galera_install recipe
```

The configuration is stored in the ```node['mysqld']['my.cnf']``` hash, and can be adapted like so


```ruby
# node['mysqld']['my.conf'][<category>][<key>] = <value>
node['mysqld']['my.cnf']['mysqld']['bind-address'] = '0.0.0.0'
```

This will expand to the following in your config file (leaving all other settings untouched)

```
[mysqld]
bind-address = 0.0.0.0
```

To remove a default option, you can pass `false` or `nil` as the value

```ruby
# Remove deprecated innodb option
default['mysqld']['my.cnf']['mysqld']['innodb_additional_mem_pool_size'] = false
```

As the configuration file is constructed from the config hash, every my.cnf configuration option is
supported.


```ruby
node['mysqld']['root_password'] = 'yourpass'
```


If you want to install percona/mysql instead of mariadb you can change the following attribute:

```ruby
node['mysqld']['db_install']
```

Valid values are `percona`, `mysql` or `mariadb`


## Recipes

### default

- Setup official MariaDB repository
- Install MariaDB server
- Configure MariaDB server according to the attributes. If no attributes are given, stick with the
  systems default


### mariadb\_repository

Sets up official MariaDB repository to install packages from.
Configure it using the following attributes

```ruby
node['mysqld']['repository']['mariadb']['version'] # Defaults to '10.1'
node['mysqld']['repository']['mariadb']['mirror']  # Defaults to HostEurope mirror
```

### percona\_repository

Sets up official Perconad repository to install packages from. This is called by default
if you set `node['mysqld']['db_install']` to percona.
Configure it using the following attributes

```ruby
node['mysqld']['repository']['percona']['version'] # Defaults to '5.7'
node['mysqld']['repository']['percona']['mirror']  # Defaults to Percona Main Mirror.
```

### install

Install mariadb/ mysql packages (according to attributes, defaults to `mariadb-server`)

### configure

Configure mysql according to attributes. Sets the databases root account (resp. debian-sys-maint on
Debian/Ubuntu systems) to use the password in `node['mysqld']['root_password']`, if the attribute is
set.

### mariadb\_galera\_init

Run `mariadb_repository` and `mariadb_galera_install` recipes, then configure as the `configure`
recipe would do, but start mariadb with `--wsrep-new-cluster --wsrep_cluster_address=gcomm://` to
initialize a new Galera cluster.

Use this if you want to setup a new Galera cluster, and run it on your first node once:

```bash
$ sudo chef-client --once -o 'recipe[mysqld::mariadb_galera_init]'
```

Once you connected the other nodes using the regular recipes, re-run chef-client as you did on the
other servers.

*Note: If you use a wrapper cookbook to configure your instances, attributes might not be available
when running the recipe with `-o recipe[]`. Create a mariadb_galera_init recipe in your wrapper
cookbook, calling this recipe if you have trouble.*


## Providers

### mysqld

You can configure your database also using the `mysqld` provider:


```ruby
include_recipe 'mysqld::mariadb_galera_install'

# Name attribute will be ignored. Choose something that makes sense for you
mysqld 'galera' do
  my_cnf { 'bind-address' => '0.0.0.0' }
end
```

### password

You can set passwords (incl. root and debian-sys-maint accounts) using this provider.
By default, the provider uses the created root/debian-sys-maint accounts depending on the system you
are on.

```ruby
mysqld_password 'root' do
  password 'get_from_data_bag_maybe?'

  # If required, you can specify your own auth-scheme here
  # auth '-u specialuser -pmypass'
end
```


# Contributing

Contributions are very welcome!

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github


# License and Authors

Authors: Chris Aumann <me@chr4.org>

License: GPLv3


CHANGELOG.md
mysqld CHANGELOG
================

This file is used to list changes made in each version of the mysqld cookbook.

2.2.0
-----

- Duplicate `my.cnf` config settings when merging with defaults, fixes Chef 13 error

2.1.0
-----

- Fix repository keys for older Ubuntu releases

2.0.0
-----

- Drop support for RHEL, pull-requests apprechiated!
  In case you are using RHEL, make sure to version-lock this cookbook to `~> 1.0.5` e.g. in your Berksfile
- Drop support for `mysql-server-5.6` (Ubuntu 14.04). Migrate to MariaDB, or version-lock this cookbook to `~> 1.0.5` e.g. in your Berksfile
- Bump default mariadb version to `10.1`
- Adapt default attributes for `mysql-5.7` and `mariadb-10.1`
- Remove workaround for deprecated Ubuntu 13.10
- Add `node['mysqld']['use_mariadb']` attribute, default to true
- Replace `mysqld::install_mariadb` and `mysqld::install_mysql` recipes with `mysqld::install``
- Rename `mysqld::mariadb_apt_repository` to `mysqld::mariadb_repository`
- Use `authentication_string` table when changing passwords

1.0.5
-----

- Update mariadb repository key

1.0.4
-----

- Fix attribute typo in `mysqld::mariadb_install` recipe

1.0.3
-----

- Use Upstart provider only on Ubuntu 13.10

1.0.2
-----

- Restart service immediately when my.cnf is updated

1.0.1
-----

- Do not force presence of `password` attribute

1.0.0
-----

- Add support for mariadb
- Add support for mariadb-galera
- Automatically set admin passwords if attribute is given (incl. modifying `/etc/mysql/debian.cnf`
  accordingly, if required)

Compatibility changes:

- Rename `node['mysqld']['packages']` attribute to `node['mysqld']['mysql_packages']`

0.3.0
-----

- Add support to remove configuration options set by default attributes

0.2.0
-----
- Do not manage `service_name`, when `service_name` is empty
- Fixes an issue with the template when the LWRP was called from another cookbook
- Use `deep_merge` to merge my.cnf hashes

0.1.0
-----
- Initial release of mysqld


attributes/defaults.rb
#
# Cookbook Name:: mysqld
# Attributes:: default
#
# Copyright 2013, Chris Aumann
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

# Attribute that defines whether MariaDB, MySQL or Percona should be used
default['mysqld']['db_install'] = 'mariadb'

# Default packages to install
default['mysqld']['mysql']['packages'] = %w(mysql-server)
default['mysqld']['mariadb']['packages'] = %w(mariadb-server)
default['mysqld']['mariadb']['galera_packages'] = %w(mariadb-galera-server)

# MariaDB repository configuration
default['mysqld']['repository']['mariadb']['version'] = '10.1'
default['mysqld']['repository']['mariadb']['mirror'] = 'http://ftp.hosteurope.de/mirror/mariadb.org/repo'

# Percona repository configuration
default['mysqld']['repository']['percona']['version'] = '5.7'
default['mysqld']['repository']['percona']['mirror'] = 'http://repo.percona.com/apt'

default['mysqld']['percona']['packages'] = %w(percona-server-server-5.7)

# Configure services
default['mysqld']['my.cnf_path'] = '/etc/mysql/my.cnf'
default['mysqld']['service_name'] = 'mysql'
default['mysqld']['includedir'] = true

# Root password, not set if nil
default['mysqld']['root_password'] = nil

# Use defaults-file for authentication (Debian)
default['mysqld']['auth'] = '--defaults-file=/etc/mysql/debian.cnf'

# Options, only set by default/ available on MariaDB
if node['mysqld']['db_install'] == 'mariadb'
  # Charset options are only set on MariaDB by default
  default['mysqld']['my.cnf']['client']['default-character-set'] = 'utf8mb4'
  default['mysqld']['my.cnf']['mysql']['default-character-set'] = 'utf8mb4'

  default['mysqld']['my.cnf']['mysqld']['character-set-server'] = 'utf8mb4'
  default['mysqld']['my.cnf']['mysqld']['collation-server'] = 'utf8mb4_general_ci'

  # This option is not present on mysql-5.7
  default['mysqld']['my.cnf']['mysqld_safe']['skip_log_error'] = true
end

# Password columns for user passwords
default['mysqld']['pwd_col'] = case node['mysqld']['db_install']
                               when 'mariadb'
                                 'Password'
                               else
                                 'authentication_string'
                               end

default['mysqld']['my.cnf']['mysqld']['user'] = 'mysql'
default['mysqld']['my.cnf']['mysqld']['port'] = 3306
default['mysqld']['my.cnf']['mysqld']['lc-messages-dir'] = '/usr/share/mysql'
default['mysqld']['my.cnf']['mysqld']['skip-external-locking'] = true
default['mysqld']['my.cnf']['mysqld']['bind-address'] = '127.0.0.1'
default['mysqld']['my.cnf']['mysqld']['key_buffer_size'] = '16M'
default['mysqld']['my.cnf']['mysqld']['max_allowed_packet'] = '16M'
default['mysqld']['my.cnf']['mysqld']['thread_stack'] = '192K'
default['mysqld']['my.cnf']['mysqld']['thread_cache_size'] = 8
default['mysqld']['my.cnf']['mysqld']['query_cache_limit'] = '1M'
default['mysqld']['my.cnf']['mysqld']['query_cache_size'] = '16M'
default['mysqld']['my.cnf']['mysqld']['pid-file'] = '/var/run/mysqld/mysqld.pid'
default['mysqld']['my.cnf']['mysqld']['socket'] = '/var/run/mysqld/mysqld.sock'
default['mysqld']['my.cnf']['mysqld']['basedir'] = '/usr'
default['mysqld']['my.cnf']['mysqld']['datadir'] = '/var/lib/mysql'
default['mysqld']['my.cnf']['mysqld']['tmpdir'] = '/tmp'
default['mysqld']['my.cnf']['mysqld']['log_error'] = '/var/log/mysql/error.log'
default['mysqld']['my.cnf']['mysqld']['expire_logs_days'] = 10
default['mysqld']['my.cnf']['mysqld']['max_binlog_size'] = '100M'
default['mysqld']['my.cnf']['mysqld']['myisam-recover-options'] = 'BACKUP'

default['mysqld']['my.cnf']['mysqld_safe']['socket'] = '/var/run/mysqld/mysqld.sock'
default['mysqld']['my.cnf']['mysqld_safe']['nice'] = 0
default['mysqld']['my.cnf']['mysqld_safe']['syslog'] = true
